tests/components/mikrotik/test_hub.py

```diff
@@ -0,0 +1,179 @@
+    ):
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+        return hass.data[mikrotik.DOMAIN][config_entry.entry_id]
+
+
+async def test_hub_setup_successful(hass):
+    """Successful setup of Mikrotik hub."""
+    with patch(
+        "homeassistant.config_entries.ConfigEntries.async_forward_entry_setup",
+        return_value=True,
+    ) as forward_entry_setup:
+        hub = await setup_mikrotik_entry(hass)
+
+    assert hub.config_entry.data == {
+        mikrotik.CONF_NAME: "Mikrotik",
+        mikrotik.CONF_HOST: "0.0.0.0",
+        mikrotik.CONF_USERNAME: "user",
+        mikrotik.CONF_PASSWORD: "pass",
+        mikrotik.CONF_PORT: 8278,
+        mikrotik.CONF_VERIFY_SSL: False,
+    }
+    assert hub.config_entry.options == {
+        mikrotik.hub.CONF_FORCE_DHCP: False,
+        mikrotik.CONF_ARP_PING: False,
+        mikrotik.CONF_DETECTION_TIME: 300,
+    }
+
+    assert hub.api.available is True
+    assert hub.signal_update == "mikrotik-update-0.0.0.0"
+    assert forward_entry_setup.mock_calls[0][1] == (hub.config_entry, "device_tracker")
+
+
+async def test_hub_setup_failed(hass):
+    """Failed setup of Mikrotik hub."""
+
+    config_entry = MockConfigEntry(domain=mikrotik.DOMAIN, data=MOCK_DATA)
+    config_entry.add_to_hass(hass)
+    # error when connection fails
+    with patch(
+        "librouteros.connect", side_effect=librouteros.exceptions.ConnectionClosed
+    ):
+
+        await hass.config_entries.async_setup(config_entry.entry_id)
+
+        assert config_entry.state == config_entries.ENTRY_STATE_SETUP_RETRY
+
+    # error when username or password is invalid
+    config_entry = MockConfigEntry(domain=mikrotik.DOMAIN, data=MOCK_DATA)
+    config_entry.add_to_hass(hass)
+    with patch(
+        "homeassistant.config_entries.ConfigEntries.async_forward_entry_setup"
+    ) as forward_entry_setup, patch(
+        "librouteros.connect",
+        side_effect=librouteros.exceptions.TrapError("invalid user name or password"),
+    ):
+
+        result = await hass.config_entries.async_setup(config_entry.entry_id)
+
+        assert result is False
+        assert len(forward_entry_setup.mock_calls) == 0
+
+
+async def test_update_failed(hass):
+    """Test failing to connect during update."""
+
+    hub = await setup_mikrotik_entry(hass)
+
+    with patch.object(
+        mikrotik.hub.MikrotikData, "command", side_effect=mikrotik.errors.CannotConnect
+    ):
+        await hub.async_update()
+
+    assert hub.api.available is False
+
+
+async def test_hub_not_support_wireless(hass):
+    """Test updating hub devices when hub doesn't support wireless interfaces."""
+
+    # test that the devices are constructed from dhcp data
+
+    hub = await setup_mikrotik_entry(hass, support_wireless=False)
+
+    assert hub.api.devices["00:00:00:00:00:01"]._params == DHCP_DATA[0]
+    assert hub.api.devices["00:00:00:00:00:01"]._wireless_params is None
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=4918093)